File: runner/redact.js
const crypto = require('crypto');

const SENSITIVE = new Set(['password', 'secret', 'token', 'authorization']);
const URI_KEYS = new Set(['uri', 'mongoUri', 'connectionString']);

function redactUri(raw) {
  const i = raw.indexOf('://');
  if (i < 0) return '[unparseable-uri]';
  const scheme = raw.slice(0, i + 3);
  const rest = raw.slice(i + 3);
  const at = rest.indexOf('@');
  if (at < 0) {
    // No credentials — but still validate that `rest` has at least a host component
    // and contains no whitespace, otherwise treat as unparseable.
    if (!rest || /\s/.test(rest)) return '[unparseable-uri]';
    return raw;
  }
  const creds = rest.slice(0, at);
  const tail = rest.slice(at);
  const colon = creds.indexOf(':');
  if (colon < 0) return raw;
  return `${scheme}${creds.slice(0, colon)}:***${tail}`;
}

function redactScript(raw) {
  const hash = crypto.createHash('sha256').update(raw).digest('hex');
  // Slice by Unicode code points (Array.from), not UTF-16 code units, so a
  // surrogate pair (e.g. an emoji) at the boundary isn't split into a lone
  // high surrogate. .length is still cheap and safe for the gate check —
  // only the slice path needs to be code-point aware.
  let head;
  if (raw.length > 200) {
    const points = Array.from(raw);
    head = (points.length > 200 ? points.slice(0, 200).join('') : raw) + '…';
  } else {
    head = raw;
  }
  return `${head} hash:${hash}`;
}

function redactCtx(ctx) {
  if (!ctx) return {};
  const out = {};
  for (const [k, v] of Object.entries(ctx)) {
    if (SENSITIVE.has(k)) out[k] = '***';
    else if (URI_KEYS.has(k) && typeof v === 'string') out[k] = redactUri(v);
    else if (k === 'script' && typeof v === 'string') out[k] = redactScript(v);
    else out[k] = v;
  }
  return out;
}

module.exports = { redactCtx, redactUri, redactScript };

File: src-tauri/src/logger/redact.rs
use super::LogCtx;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};

const SENSITIVE_KEYS: &[&str] = &["password", "secret", "token", "authorization"];
const URI_KEYS: &[&str] = &["uri", "mongoUri", "connectionString"];

fn redact_uri(raw: &str) -> String {
    // mongodb://user:password@host/db → mongodb://user:***@host/db
    // Parse manually — MongoDB URIs include `+srv` variants that url::Url handles fine,
    // but we use a substring approach to keep the dependency footprint small.
    if let Some(scheme_end) = raw.find("://") {
        let (scheme, rest) = raw.split_at(scheme_end + 3);
        if let Some(at) = rest.find('@') {
            let creds = &rest[..at];
            let tail = &rest[at..];
            if let Some(colon) = creds.find(':') {
                let user = &creds[..colon];
                return format!("{scheme}{user}:***{tail}");
            }
        }
        return raw.to_owned();
    }
    "[unparseable-uri]".to_owned()
}

fn redact_script(raw: &str) -> String {
    let mut h = Sha256::new();
    h.update(raw.as_bytes());
    let hash = hex::encode(h.finalize());
    let head = if raw.chars().count() > 200 {
        let truncated: String = raw.chars().take(200).collect();
        format!("{truncated}…")
    } else {
        raw.to_owned()
    };
    format!("{head} hash:{hash}")
}

pub fn redact_ctx(ctx: LogCtx) -> LogCtx {
    let mut out = LogCtx::new();
    for (k, v) in ctx {
        if SENSITIVE_KEYS.contains(&k.as_str()) {
            out.insert(k, json!("***"));
        } else if URI_KEYS.contains(&k.as_str()) {
            if let Value::String(s) = &v {
                out.insert(k, json!(redact_uri(s)));
            } else {
                out.insert(k, v);
            }
        } else if k == "script" {
            if let Value::String(s) = &v {
                out.insert(k, json!(redact_script(s)));
            } else {
                out.insert(k, v);
            }
        } else {
            out.insert(k, v);
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::logctx;

    #[test]
    fn redacts_mongo_uri_password() {
        let r = redact_ctx(logctx! { "uri" => "mongodb://user:secret@host/db" });
        assert_eq!(r.get("uri").unwrap(), &json!("mongodb://user:***@host/db"));
    }

    #[test]
    fn masks_password_and_token_fields() {
        let r = redact_ctx(logctx! { "password" => "p", "token" => "t" });
        assert_eq!(r.get("password").unwrap(), &json!("***"));
        assert_eq!(r.get("token").unwrap(), &json!("***"));
    }

    #[test]
    fn truncates_and_hashes_script() {
        let script: String = "a".repeat(500);
        let r = redact_ctx(logctx! { "script" => script });
        let out = r.get("script").unwrap().as_str().unwrap().to_owned();
        assert!(out.contains("…"));
        assert!(out.contains("hash:"));
    }

    #[test]
    fn passes_through_unrelated_fields() {
        let r = redact_ctx(logctx! { "connId" => "c_1", "page" => 3 });
        assert_eq!(r.get("connId").unwrap(), &json!("c_1"));
        assert_eq!(r.get("page").unwrap(), &json!(3));
    }
}
